src/io/memory.cr

```diff
@@ -458,6 +450,28 @@ class IO::Memory < IO
     end
   end
 
+  private def increase_capacity_by(count)
+    raise IO::EOFError.new if count >= Int32::MAX - bytesize
+
+    new_bytesize = @pos + count
+    return if new_bytesize <= @capacity
+
+    check_resizeable
+
+    new_capacity = calculate_new_capacity(new_bytesize)
+    resize_to_capacity(new_capacity)
+  end
+
+  private def calculate_new_capacity(new_bytesize : Int32)
+    # If the new bytesize is bigger than 1 << 30, the next power of two would
+    # be 1 << 31, which is out of range for Int32.
+    # So we limit the capacity to Int32::MAX in order to be able to use the
+    # range (1 << 30) < new_bytesize < Int32::MAX
+    return Int32::MAX if new_bytesize > 1 << 30
+
+    Math.pw2ceil(new_bytesize)
```

Comment:
Unrelated, but I wonder if we shouldn't change the formula like we did for [array](https://github.com/crystal-lang/crystal/pull/11482)

Reply:
And the converse is also true; that we should cap the max of array too

Reply:
Maybe time for a `Growable` module...

Reply:
=> #14173